File: NodeJS/crawler_hasdata.js
const axios = require('axios');
const fs = require('fs/promises');

const API_KEY = 'YOUR-API-KEY';

const headers = {
  'x-api-key': API_KEY,
  'Content-Type': 'application/json'
};

const payload = {
  limit: 20,
  urls: ['https://demo.nopcommerce.com'],
  outputFormat: ['json']
};

async function delay(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

(async () => {
  try {
    const startRes = await axios.post(
      'https://api.hasdata.com/scrapers/crawler/jobs',
      payload,
      { headers }
    );

    const jobId = startRes.data.id;
    console.log(`Started job with ID: ${jobId}`);

    let status;
    let jobData;
    while (true) {
      const statusRes = await axios.get(
        `https://api.hasdata.com/scrapers/jobs/${jobId}`,
        { headers }
      );

      jobData = statusRes.data;
      status = jobData.status;
      console.log(`Job status: ${status}`);

      if (['finished', 'failed', 'cancelled'].includes(status)) {
        break;
      }

      await delay(10000);
    }

    if (status !== 'finished') {
      console.error(`Job ended with status: ${status}`);
      return;
    }

    const jsonUrl = jobData.data.json;
    const outputPath = `results_${jobId}.json`;

    const jsonRes = await axios.get(jsonUrl);
    const rawData = jsonRes.data;

    const urls = rawData
      .filter(entry => entry.url)
      .map(entry => entry.url);

    await fs.writeFile(outputPath, JSON.stringify(urls, null, 2), 'utf-8');
    console.log(`Saved to ${outputPath}`);
  } catch (err) {
    console.error('Error:', err.message);
  }
})();

File: Python/crawler_hasdata.py
import time
import requests
import json

API_KEY = 'YOUR-API-KEY'

headers = {
    'x-api-key': API_KEY,
    'Content-Type': 'application/json'
}

payload = {
    "limit": 20,
    "urls": ["https://demo.nopcommerce.com"],
    "outputFormat": ["json"]
}

response = requests.post(
    "https://api.hasdata.com/scrapers/crawler/jobs",
    json=payload,
    headers=headers
)
response.raise_for_status()
job_id = response.json().get("id")
print(f"Started job with ID: {job_id}")

while True:
    response = requests.get(
        f"https://api.hasdata.com/scrapers/jobs/{job_id}",
        headers=headers
    )
    data = response.json()
    status = data.get("status")
    print(f"Job status: {status}")

    if status in ["finished", "failed", "cancelled"]:
        break

    time.sleep(10)

json_url = data["data"]["json"]
output_path = f"results_{job_id}.json"

response = requests.get(json_url)
response.raise_for_status()
raw_data = response.json()

urls = [entry["url"] for entry in raw_data if "url" in entry]

with open(output_path, "w", encoding="utf-8") as f:
    json.dump(urls, f, indent=2)

print(f"Saved to {output_path}")
